cni/pkg/taint/taint.go

```diff
@@ -0,0 +1,387 @@
+}
+
+// taint node with specific taint name with effect of no schedule
+//do nothing if it already have the readiness taint
+func (ts TaintSetter) AddReadinessTaint(node *v1.Node) error {
+	var taintsExists bool
+	for _, taint := range node.Spec.Taints {
+		if taint.Key == TaintName {
+			taintsExists = true
+			break
+		}
+	}
+	if taintsExists {
+		log.Debugf("%v already present on node %v", TaintName, node.Name)
+		return nil
+	}
+	node.Spec.Taints = append(node.Spec.Taints, v1.Taint{
+		Key:    TaintName,
+		Effect: v1.TaintEffectNoSchedule,
+	})
+	updatedNodeWithTaint, err := ts.Client.CoreV1().Nodes().Update(context.TODO(), node, metav1.UpdateOptions{})
+	if err != nil || updatedNodeWithTaint == nil {
+		return fmt.Errorf("failed to update node %v after adding taint: %v", node.Name, err)
+	}
+	log.Infof("Successfully added taint on node %v", updatedNodeWithTaint.Name)
+	return nil
+}
+
+//get all critical pods in a node
+func (ts TaintSetter) GetAllCriticalPodsInNode(node *v1.Node) (podlist v1.PodList, err error) {
+	candidates, err := ts.ListCandidatePods()
+	if err != nil {
+		return podlist, err
+	}
+	podlist.Items = []v1.Pod{}
+	for _, pod := range candidates.Items {
+		if pod.Spec.NodeName == node.Name {
+			podlist.Items = append(podlist.Items, pod)
+		}
+	}
+	return podlist, nil
+}
+
+//return true if labels contain at least one selector in labelselector class
+func containsLabel(labels map[string]string, selector string) bool {
+	labelset := make(map[string]struct{})
+	for key, val := range labels {
+		labelset[fmt.Sprintf("%s=%s", key, val)] = struct{}{}
+	}
+	if _, ok := labelset[selector]; ok {
+		return true
+	}
+	return false
+}
+
+//get pod's node and return error if it cannot find node by pod's node name
+func (ts TaintSetter) GetNodeByPod(pod *v1.Pod) (*v1.Node, error) {
+	updatedNode, err := ts.Client.CoreV1().Nodes().Get(context.TODO(), pod.Spec.NodeName, metav1.GetOptions{})
+	if err != nil || updatedNode == nil {
+		return nil, fmt.Errorf("failed to get node %v for pod %v :  %v", pod.Spec.NodeName, pod.Name, err)
+	}
+	return updatedNode, nil
+}
+
+// DeleteTaint removes all the the taints that have the same key and effect to given taintToDelete.
+func deleteTaint(taints []v1.Taint, taintToDelete *v1.Taint) []v1.Taint {
+	newTaints := []v1.Taint{}
+	for i := range taints {
+		if taintToDelete.MatchTaint(&taints[i]) {
+			continue
+		}
+		newTaints = append(newTaints, taints[i])
+	}
+	return newTaints
+}
+
+//check the readiness of node by looping over all critical pod inside the node
+//if all labels selector in each namespace get satisfied, return true
+func (ts TaintSetter) CheckNodeReadiness(node *v1.Node) bool {
+	if ts.configs == nil || len(ts.configs) == 0 {
+		return true
+	}
+	podList, err := ts.GetAllCriticalPodsInNode(node)
+	if err != nil {
+		log.Infof("error happened in get critical pod, %v", err)
+		return false
+	}
+	nsToPods := ts.buildNamespaceToPodsMap(podList)
+	nsToLables := ts.buildLabelReadiness()
```

Comment:
This also seems unnecessary. I think this function needs a rewrite -- the logic should be something like this (very much pseudocode):

```
for criticalPodDefinition in criticalLabels:
   if len(ListCandidatePods(node.Name, criticalPodDefinition.podNamespace, criticalPodDefinition.LabelSelector)) > 0:
     return true
```